docs/source/futures.rst

```diff
@@ -0,0 +1,522 @@
+will handle this automatically.
+
+.. code-block:: python
+
+   c = client.submit(add, a, b)  # calls add on the results of a and b
+
+You can use ``map`` to call ``submit`` on the same function and many inputs:
+
+.. code-block:: python
+
+   futures = client.map(inc, range(1000))
+
+However note that each task comes with about 1ms of overhead.  If you want to
+map a function over a large number of inputs then you might consider
+:doc:`dask.bag <bag>` or :doc:`dask.dataframe <dataframe>` instead.
+
+Move Data
+---------
+
+.. autosummary::
+   Future.result
+   Client.gather
+   Client.scatter
+
+Given any future you can call the ``.result`` method to gather the result.
+This will block until the future is done computing and then transfer the result
+back to your local process if necessary.
+
+.. code-block:: python
+
+   >>> c.result()
+   32
+
+You can gather many results concurrently using the ``Client.gather`` method.
+This can be more efficient than calling ``.result()`` on each future
+sequentially.
+
+.. code-block:: python
+
+   >>> # results = [future.result() for future in futures]
+   >>> results = client.gather(futures)  # this can be faster
+
+If you have important local data that you want to include in your computation
+you can either include it as a normal input to a submit or map call:
+
+.. code-block:: python
+
+   >>> df = pd.read_csv('training-data.csv')
+   >>> future = client.submit(my_function, df)
+
+Or you can ``scatter`` it explicitly.  Scattering moves your data to a worker
+and gives you back a future pointing to that data:
+
+.. code-block:: python
+
+   >>> remote_df = client.scatter(df)
+   >>> remote_df
+   <Future: status: finished, type: DataFrame, key: bbd0ca93589c56ea14af49cba470006e>
+
+   >>> future = client.submit(my_function, remote_df)
+
+Both of these accomplish the same result, but using scatter can sometimes be
+faster, especially if you are using processes or distributed workers (where
+data transfer is necessary) and you want to use ``df`` in many computations.
+Scattering the data beforehand avoids excessive data movement.
+
+Scattering lists scatters all elements individually.  Dask will spread these
+elements evenly throughout workers in a round-robin fashion:
+
+.. code-block:: python
+
+   >>> client.scatter([1, 2, 3])
+   [<Future: status: finished, type: int, key: c0a8a20f903a4915b94db8de3ea63195>,
+    <Future: status: finished, type: int, key: 58e78e1b34eb49a68c65b54815d1b158>,
+    <Future: status: finished, type: int, key: d3395e15f605bc35ab1bac6341a285e2>]
+
+References, Cancellation, and Exceptions
+----------------------------------------
+
+.. autosummary::
+   Future.cancel
+   Future.exception
+   Future.traceback
+   Client.cancel
+
+Dask will only compute and hold onto results for which there are active
+futures.  In this way your local variables define what is active in Dask.  If
+you delete the last future to a piece of remote data then Dask will feel free
```

Comment:
"the last" is a bit confusing to me - in what situations would there be more than one future pointing to the same data? Perhaps

```
When a `Future` is garbage collected by python, `Dask will feel free ...
```